File: go/evals/runner.go
package evals

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

type EvalCase struct {
	Name         string
	Description  string
	SystemPrompt string
	UserMessage  string
	Tools        []ToolDefinition
	ExpectedTool string
	MaxTurns     int
}

type EvalResult struct {
	CaseName   string
	ModelName  string
	Success    bool
	Turns      int
	ToolCalls  []map[string]interface{}
	Error      string
	DurationMs int64
}

type EvalReport struct {
	Timestamp string
	Models    []string
	Results   []EvalResult
}

func (r *EvalReport) PassRate(model string) float64 {
	var passed, total int
	for _, res := range r.Results {
		if res.ModelName == model {
			total++
			if res.Success {
				passed++
			}
		}
	}
	if total == 0 {
		return 0.0
	}
	return float64(passed) / float64(total)
}

type EvalRunner struct {
	BaseURL  string
	MaxTurns int
	Results  []EvalResult
}

func NewEvalRunner(baseURL string, maxTurns int) *EvalRunner {
	return &EvalRunner{
		BaseURL:  baseURL,
		MaxTurns: maxTurns,
	}
}

type ToolExecutor func(toolName string, args map[string]interface{}) string

func (r *EvalRunner) RunCase(case_ EvalCase, model string, toolExecutor ToolExecutor) EvalResult {
	start := time.Now()
	client := NewModelClient(r.BaseURL, model, "")

	messages := []ChatMessage{
		{Role: "system", Content: case_.SystemPrompt},
		{Role: "user", Content: case_.UserMessage},
	}

	var toolCallsMade []map[string]interface{}
	turns := 0
	errMsg := ""

	for turns < case_.MaxTurns {
		tools := case_.Tools
		result, err := client.Complete(messages, tools, 0.0, 2048)
		turns++
		if err != nil {
			errMsg = err.Error()
			break
		}

		if len(result.ToolCalls) == 0 {
			break
		}

		for _, tc := range result.ToolCalls {
			toolCallsMade = append(toolCallsMade, map[string]interface{}{
				"turn":      turns,
				"name":      tc.Name,
				"arguments": tc.Arguments,
			})

			if tc.Name == case_.ExpectedTool {
				durationMs := time.Since(start).Milliseconds()
				return EvalResult{
					CaseName:   case_.Name,
					ModelName:  model,
					Success:    true,
					Turns:      turns,
					ToolCalls:  toolCallsMade,
					DurationMs: durationMs,
				}
			}

			var args map[string]interface{}
			_ = json.Unmarshal([]byte(tc.Arguments), &args)

			toolResult := toolExecutor(tc.Name, args)

			messages = append(messages, ChatMessage{
				Role:    "assistant",
				Content: "",
			})
			messages = append(messages, ChatMessage{
				Role:    "tool",
				Content: toolResult,
			})
		}
	}

	durationMs := time.Since(start).Milliseconds()
	if errMsg == "" {
		errMsg = fmt.Sprintf("Expected tool '%s' not called in %d turns", case_.ExpectedTool, turns)
	}
	return EvalResult{
		CaseName:   case_.Name,
		ModelName:  model,
		Success:    false,
		Turns:      turns,
		ToolCalls:  toolCallsMade,
		Error:      errMsg,
		DurationMs: durationMs,
	}
}

func (r *EvalRunner) RunEvals(cases_ []EvalCase, models []string, toolExecutor ToolExecutor) EvalReport {
	report := EvalReport{
		Timestamp: time.Now().Format(time.RFC3339),
		Models:    models,
	}

	for _, model := range models {
		fmt.Printf("\n=== Testing model: %s ===\n", model)
		for _, case_ := range cases_ {
			fmt.Printf("  Running: %s...", case_.Name)
			result := r.RunCase(case_, model, toolExecutor)
			r.Results = append(r.Results, result)
			report.Results = append(report.Results, result)

			status := "PASS"
			if !result.Success {
				status = "FAIL"
			}
			fmt.Printf("%s (%d turns, %dms)\n", status, result.Turns, result.DurationMs)

			if !result.Success {
				fmt.Printf("    Error: %s\n", result.Error)
			}
		}
	}

	return report
}

func (r *EvalRunner) SaveReport(report EvalReport, outputPath string) {
	dir := filepath.Dir(outputPath)
	_ = os.MkdirAll(dir, 0755)

	output := map[string]interface{}{
		"timestamp": report.Timestamp,
		"models":    report.Models,
		"results":   []map[string]interface{}{
			// Results will be serialized individually below
		},
	}

	results := make([]map[string]interface{}, len(report.Results))
	for i, res := range report.Results {
		results[i] = map[string]interface{}{
			"case":        res.CaseName,
			"model":       res.ModelName,
			"success":     res.Success,
			"turns":       res.Turns,
			"tool_calls":  res.ToolCalls,
			"error":       res.Error,
			"duration_ms": res.DurationMs,
		}
	}
	output["results"] = results

	data, _ := json.MarshalIndent(output, "", "  ")
	if err := os.WriteFile(outputPath, data, 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Warning: failed to write report: %v\n", err)
	}
}

File: typescript/src/evals/runner.ts
import { ModelClient, ChatMessage, ToolDefinition } from "./models";

export { ToolDefinition };

export interface EvalCase {
  name: string;
  description: string;
  systemPrompt: string;
  userMessage: string;
  tools: ToolDefinition[];
  expectedTool: string;
  maxTurns: number;
}

export interface EvalResult {
  case_name: string;
  model_name: string;
  success: boolean;
  turns: number;
  tool_calls: Array<{ turn: number; name: string; arguments: string }>;
  error: string;
  duration_ms: number;
}

export interface EvalReport {
  timestamp: string;
  models: string[];
  results: EvalResult[];
}

export class EvalRunner {
  private baseURL: string;
  private maxTurns: number;
  private results: EvalResult[] = [];

  constructor(baseURL: string, maxTurns: number = 10) {
    this.baseURL = baseURL;
    this.maxTurns = maxTurns;
  }

  passRate(model: string): number {
    const modelResults = this.results.filter((r) => r.model_name === model);
    if (modelResults.length === 0) return 0.0;
    const passed = modelResults.filter((r) => r.success).length;
    return passed / modelResults.length;
  }

  async runCase(
    case_: EvalCase,
    model: string,
    toolExecutor: (toolName: string, args: Record<string, unknown>) => Promise<string>
  ): Promise<EvalResult> {
    const start = Date.now();
    const client = new ModelClient(this.baseURL, model);

    const messages: ChatMessage[] = [
      { role: "system", content: case_.systemPrompt },
      { role: "user", content: case_.userMessage },
    ];

    const toolCallsMade: Array<{ turn: number; name: string; arguments: string }> = [];
    let turns = 0;
    let error = "";

    try {
      while (turns < case_.maxTurns) {
        const result = await client.complete(messages, case_.tools, 0.0, 2048);
        turns++;

        if (result.tool_calls.length === 0) {
          break;
        }

        for (const tc of result.tool_calls) {
          toolCallsMade.push({
            turn: turns,
            name: tc.name,
            arguments: tc.arguments,
          });

          if (tc.name === case_.expectedTool) {
            const durationMs = Date.now() - start;
            return {
              case_name: case_.name,
              model_name: model,
              success: true,
              turns: turns,
              tool_calls: toolCallsMade,
              error: "",
              duration_ms: durationMs,
            };
          }

          const args = JSON.parse(tc.arguments);
          const toolResult = await toolExecutor(tc.name, args);

          messages.push({
            role: "assistant",
            content: "",
          });
          messages.push({
            role: "tool",
            content: toolResult,
          });
        }
      }

      const durationMs = Date.now() - start;
      error = `Expected tool '${case_.expectedTool}' not called in ${turns} turns`;
      return {
        case_name: case_.name,
        model_name: model,
        success: false,
        turns: turns,
        tool_calls: toolCallsMade,
        error: error,
        duration_ms: durationMs,
      };
    } catch (e) {
      const durationMs = Date.now() - start;
      return {
        case_name: case_.name,
        model_name: model,
        success: false,
        turns: turns,
        tool_calls: toolCallsMade,
        error: e instanceof Error ? e.message : String(e),
        duration_ms: durationMs,
      };
    }
  }

  async runEvals(
    cases: EvalCase[],
    models: string[],
    toolExecutor: (toolName: string, args: Record<string, unknown>) => Promise<string>
  ): Promise<EvalReport> {
    const report: EvalReport = {
      timestamp: new Date().toISOString(),
      models: models,
      results: [],
    };

    for (const model of models) {
      console.log(`\n=== Testing model: ${model} ===`);
      for (const case_ of cases) {
        process.stdout.write(`  Running: ${case_.name}...`);
        const result = await this.runCase(case_, model, toolExecutor);
        this.results.push(result);
        report.results.push(result);

        const status = result.success ? "PASS" : "FAIL";
        console.log(`${status} (${result.turns} turns, ${result.duration_ms}ms)`);

        if (!result.success) {
          console.log(`    Error: ${result.error}`);
        }
      }
    }

    return report;
  }

  saveReport(report: EvalReport, outputPath: string): void {
    const fs = require("fs");
    const dir = require("path").dirname(outputPath);
    fs.mkdirSync(dir, { recursive: true });

    const output = {
      timestamp: report.timestamp,
      models: report.models,
      results: report.results.map((r) => ({
        case: r.case_name,
        model: r.model_name,
        success: r.success,
        turns: r.turns,
        tool_calls: r.tool_calls,
        error: r.error,
        duration_ms: r.duration_ms,
      })),
    };

    fs.writeFileSync(outputPath, JSON.stringify(output, null, 2));
  }
}